File: logmate-client-python/src/logmate/transport.py
from __future__ import annotations

import json
import socket
import urllib.error
import urllib.request
from typing import Any, Mapping

from .errors import (
    _RequestFailure,
    _friendly_error,
    _http_error_message,
    _is_retryable_http_status,
)


class TransportMixin:
    def _post(
        self,
        path: str,
        payload: Mapping[str, Any],
        *,
        origin_instance_id: str = "",
    ) -> dict[str, Any]:
        headers = {
            "Accept": "application/json",
            "Content-Type": "application/json",
        }
        if self.instance_id:
            headers["X-Sender-Instance-ID"] = self.instance_id
            headers["X-Sender-Instance-Token"] = self.instance_token
        if origin_instance_id and origin_instance_id != self.instance_id:
            headers["X-LogMate-Origin-Instance-ID"] = origin_instance_id

        try:
            encoded_payload = json.dumps(
                payload,
                ensure_ascii=False,
                default=str,
            ).encode("utf-8")
        except Exception as error:
            raise _RequestFailure(
                f"could not convert the log to JSON: {error}",
                retryable=False,
            ) from None

        request = urllib.request.Request(
            self.api_url + path,
            encoded_payload,
            headers,
            method="POST",
        )

        try:
            with urllib.request.urlopen(request, timeout=float(self.timeout)) as response:
                body = response.read()
        except urllib.error.HTTPError as error:
            raise _RequestFailure(
                _http_error_message(error),
                retryable=_is_retryable_http_status(error.code),
            ) from None
        except urllib.error.URLError as error:
            raise _RequestFailure(_friendly_error(error), retryable=True) from None
        except (
            TimeoutError,
            socket.timeout,
            ConnectionRefusedError,
            ConnectionResetError,
            ConnectionAbortedError,
            BrokenPipeError,
            socket.gaierror,
            OSError,
        ) as error:
            raise _RequestFailure(_friendly_error(error), retryable=True) from None
        except Exception as error:
            raise _RequestFailure(_friendly_error(error), retryable=False) from None

        if not body:
            return {}

        try:
            decoded = body.decode("utf-8")
            parsed = json.loads(decoded)
        except Exception as error:
            raise _RequestFailure(_friendly_error(error), retryable=False) from None

        if not isinstance(parsed, dict):
            raise _RequestFailure(
                "the API returned JSON, but the content is not an object.",
                retryable=False,
            )
        return parsed

File: logmate-client-ts/src/transport.ts
import { RequestFailure, friendlyError, httpError } from "./errors.js";

export class Transport {
  constructor(
    private readonly apiUrl: string,
    private readonly timeout: number,
    private readonly credentials: () => { instanceId: string; instanceToken: string },
  ) {}

  async post(
    path: string,
    payload: Record<string, unknown>,
    originInstanceId = "",
  ): Promise<Record<string, unknown>> {
    const { instanceId, instanceToken } = this.credentials();
    const headers: Record<string, string> = {
      Accept: "application/json",
      "Content-Type": "application/json",
    };
    if (instanceId) {
      headers["X-Sender-Instance-ID"] = instanceId;
      headers["X-Sender-Instance-Token"] = instanceToken;
    }
    if (originInstanceId && originInstanceId !== instanceId) {
      headers["X-LogMate-Origin-Instance-ID"] = originInstanceId;
    }

    let body: string;
    try {
      body = JSON.stringify(payload);
    } catch (error) {
      throw new RequestFailure(`could not convert the log to JSON: ${friendlyError(error)}`, false);
    }

    let response: Response;
    try {
      response = await fetch(this.apiUrl + path, {
        method: "POST",
        headers,
        body,
        signal: AbortSignal.timeout(this.timeout),
      });
    } catch (error) {
      throw new RequestFailure(friendlyError(error), true);
    }
    if (!response.ok) throw await httpError(response);
    const text = await response.text();
    if (!text) return {};
    try {
      const parsed: unknown = JSON.parse(text);
      if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
        throw new RequestFailure("the API returned JSON, but the content is not an object.", false);
      }
      return parsed as Record<string, unknown>;
    } catch (error) {
      if (error instanceof RequestFailure) throw error;
      throw new RequestFailure("the API returned a response that is not valid JSON.", false);
    }
  }
}
